apollo-router/tests/integration/subgraph_mocks_plugin/mod.rs

```diff
@@ -0,0 +1,270 @@
+                let response = response.body(body).unwrap();
+                Ok(new_subgraph_response(request, response))
+            }
+        })
+        .boxed()
+    }
+}
+
+fn subgraph_call(
+    config: &SubgraphConfig,
+    subgraph_schema: &Valid<Schema>,
+    request: &graphql::Request,
+) -> Result<graphql::Response, Vec<GraphQLError>> {
+    let query = request.query.as_deref().unwrap_or("");
+    let doc = ExecutableDocument::parse_and_validate(subgraph_schema, query, "query")
+        .map_err(|e| e.errors.iter().map(|e| e.to_json()).collect::<Vec<_>>())?;
+    let operation = doc
+        .operations
+        .get(request.operation_name.as_deref())
+        .map_err(|e| vec![e.to_graphql_error(&doc.sources)])?;
+    let variable_values = coerce_variable_values(subgraph_schema, operation, &request.variables)
+        .map_err(|e| vec![e.to_graphql_error(&doc.sources)])?;
+    let object_type_name = operation.object_type();
+    let plain_error = |message: &str| vec![GraphQLError::new(message, None, &doc.sources)];
+    let root_operation_object_type_def = subgraph_schema
+        .get_object(object_type_name)
+        .ok_or_else(|| plain_error("undefined root operation object type"))?;
+    let (mode, root_mocks) = match operation.operation_type {
+        OperationType::Query => (execution::engine::ExecutionMode::Normal, &config.query),
+        OperationType::Mutation => (
+            execution::engine::ExecutionMode::Sequential,
+            config
+                .mutation
+                .as_ref()
+                .ok_or_else(|| plain_error("mutation is not supported"))?,
+        ),
+        OperationType::Subscription => return Err(plain_error("subscription not supported")),
+    };
+    let initial_value = RootResolver {
+        root_mocks,
+        entities: &config.entities,
+    };
+    let mut errors = Vec::new();
+    let path = None;
+    let data = match execution::engine::execute_selection_set(
+        subgraph_schema,
+        &doc,
+        &variable_values,
+        &mut errors,
+        path,
+        mode,
+        root_operation_object_type_def,
+        &initial_value,
+        &operation.selection_set.selections,
+    ) {
+        Ok(map) => JsonValue::Object(map),
+        Err(execution::engine::PropagateNull) => JsonValue::Null,
+    };
+    Ok(graphql::Response::builder()
+        .data(data)
+        .errors(errors.into_iter().map(Into::into).collect())
+        .extensions(JsonMap::new())
+        .build())
+}
+
+struct RootResolver<'a> {
+    root_mocks: &'a JsonMap,
+    entities: &'a [JsonMap],
+}
+
+struct MockResolver<'a> {
+    mocks: &'a JsonMap,
+}
+
+impl<'a> RootResolver<'a> {
+    fn find_entities(&self, representation: &JsonMap) -> Option<&'a JsonMap> {
+        self.entities.iter().find(|entity| {
+            representation
+                .iter()
+                .all(|(k, v)| entity.get(k).is_some_and(|value| value == v))
```

Comment:
This linear search would not be appropriate for a production server with many entities, but mocks typically have very few